### PR TITLE
ARP metric added

### PR DESCRIPTION
**Pull request by mikesokolovv**
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Optimization

### CODE REVIEW

File: tests/metrics/test_popularity.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    def test_correct_arp_values(
+        self, recommendations: pd.DataFrame, interactions: pd.DataFrame, k: int, expected: pd.Series
+    ) -> None:
+        arp = ARP(k)
+
+        actual = arp.calc_per_user(recommendations, interactions)
+        pd.testing.assert_series_equal(actual, expected, check_names=False)
+
+        actual_mean = arp.calc(recommendations, interactions)
+        assert actual_mean == expected.mean()
```
Comment (feldlime) [inline tests/metrics/test_popularity.py:63]: Please add tests for edge cases
- when there is no data
- when there is an item in recommendations that is not present in interactions (I suppose the code will fail here)

**Review by feldlime — CHANGES_REQUESTED** [review]

File: rectools/metrics/popularity.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import attr
+import pandas as pd
+
+from rectools import Columns
+from rectools.metrics.base import MetricAtK
+from rectools.utils import select_by_type
+
+
+@attr.s
+class ARP(MetricAtK):
```
[the remark this answers is not in the document]
Comment (blondered) [reply rectools/metrics/popularity.py:29]: I agree that it's not a common name. Let's keep it up with "MeanInvUserFreq" logic. "AvgRecPopularity"?

[the remark this answers is not in the document]
Comment (feldlime) [reply rectools/metrics/popularity.py:29]: Sounds much clearer for me

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 - Methods for conversion `Interactions` to raw form and for getting raw interactions from `Dataset` ([#69](https://github.com/MobileTeleSystems/RecTools/pull/69))
+- `ARP (Average Recommendation Popularity)` to `metrics`
```
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply CHANGELOG.md]: Done.

File: rectools/metrics/popularity.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Metric name changed to "AvgRecPopularity".

File: rectools/metrics/popularity.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        prev_interactions : pd.DataFrame
+            Table with previous user-item interactions,
+            with columns `Columns.User`, `Columns.Item`.
+
+        Returns
+        -------
+        dict(int->int)
+            Set with items' popularity rating (key - item id, value - number of interactions with item in training set).
+        """
+        return Counter(prev_interactions[Columns.Item])
```
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Counter was changed to `.value_counts()`.

File: rectools/metrics/popularity.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            Table with previous user-item interactions,
+            with columns `Columns.User`, `Columns.Item`.
+
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        pop_items = self.get_pop(prev_interactions)
+        arp = reco.groupby(Columns.User).apply(
```
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: `Apply` was changed to `agg`, added check on new (unseen in interactions set) items in recommendations dataframe.

File: tests/metrics/test_popularity.py (hunk 1 above)
Comment (mikesokolovv) [reply tests/metrics/test_popularity.py:64]: Tests were added.

File: rectools/metrics/popularity.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        pop_items = self.get_pop(prev_interactions)
+        reco_prepared = reco.query(f"{Columns.Rank} <= @self.k")
+        arp = reco_prepared.groupby(Columns.User)[Columns.Item].agg(
+            lambda x: sum(pop_items[i] if i in pop_items else 0 for i in x) / x.nunique()
```
Comment (blondered) [inline rectools/metrics/popularity.py:112]: You can still make it much faster. Forget if and lambda. If you switch to joining dataframes with necessary data you will make all computations in vector form. And save a lot of time. For example:

`reco = reco.join(items_popularity, on="item_id", how="left")`
`reco["popularity"] = reco["popularity"].fillna(0) / k`

It is already divided by k. So afterwards you can just groupby by user and sum.

File: rectools/metrics/popularity.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        )
+        return arp
+
+    def get_pop(self, prev_interactions: pd.DataFrame) -> pd.Series:
```
Comment (blondered) [inline rectools/metrics/popularity.py:116]: This method does only one thing. So there is no need for a separate func. Just do value_counts() in the main func.

**Review by blondered — CHANGES_REQUESTED** [review]

File: rectools/metrics/popularity.py (hunk 7 above)
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: `get_pop` method was deleted.

File: rectools/metrics/popularity.py (hunk 6 above)
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Computations were changed to vector form.

File: rectools/metrics/popularity.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        item_popularity = prev_interactions[Columns.Item].value_counts()
+        item_popularity.name = "popularity"
+
+        reco_k = reco.query(f"{Columns.Rank} <= @self.k")
+        reco_max_k = reco_k.groupby(Columns.User)[Columns.Rank].transform("count")
```
Comment (feldlime) [inline rectools/metrics/popularity.py:113]: Let's please make the code and the variable name match each other: if it's `reco_MAX_k`, then it should be `max` aggregate function, or if you use `count`, please use an appropriate name

File: rectools/metrics/popularity.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        reco_prepared = reco_k.join(item_popularity, on=Columns.Item, how="left")
+        reco_prepared["popularity"] = reco_prepared["popularity"].fillna(0) / reco_max_k
```
Comment (feldlime) [inline rectools/metrics/popularity.py:115]: `pandas` docs don't say that order after `join` is preserved, so we can't rely on it

It's better to add a column with the number of reco to this dataframe

Comment (feldlime) [reply rectools/metrics/popularity.py:115]: Btw, it's more efficient to do actions in the next order:
1. join reco to `item_popularity`
2. `fillna`
3. group by user and calculate simultaneously 1) sum of popularity, 2) number of reco
4. divide 

It's more efficient because you don't need to store `reco_max_k` for long dataframe with useless repeated values, and also because you execute division only for small grouped data

File: rectools/metrics/popularity.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    -------
+    dict(str->float)
+        Dictionary where keys are the same as keys in `metrics`
+        and values are metric calculation results.
+    """
+    results = {}
+
+    # ARP
+    pop_metrics: tp.Dict[str, AvgRecPopularity] = select_by_type(metrics, AvgRecPopularity)
+    if pop_metrics:
```
Comment (feldlime) [inline rectools/metrics/popularity.py:156]: useless `if`

**Review by feldlime — CHANGES_REQUESTED** [review]

File: rectools/metrics/popularity.py (hunk 8 above)
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: `reco_max_k` was deleted due to calculations logic change to usage of `agg`.

File: rectools/metrics/popularity.py (hunk 9 above)
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Calculations logic was changed to `agg`. It's a bit more time consuming, but no additional memory is used and code looks uncluttered. I guess it's a fair trade-off here.

File: rectools/metrics/popularity.py (hunk 10 above)
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: `if` was deleted.

File: rectools/metrics/popularity.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        item_popularity = prev_interactions[Columns.Item].value_counts()
+        item_popularity.name = "popularity"
+
+        reco_k = reco.query(f"{Columns.Rank} <= @self.k")
+        reco_prepared = reco_k.join(item_popularity, on=Columns.Item, how="left").fillna(0)
```
Comment (feldlime) [inline rectools/metrics/popularity.py:113]: Let's fill nans only in one column
image

**Review by feldlime — CHANGES_REQUESTED** [review]

File: rectools/metrics/popularity.py (hunk 9 above)
Comment (feldlime) [reply rectools/metrics/popularity.py]: Let me disagree with this

It's much more time consuming, and additional memory consuming is really small

image

**Review by blondered — CHANGES_REQUESTED** [review]

File: rectools/metrics/popularity.py (hunk 9 above)
Comment (feldlime) [reply rectools/metrics/popularity.py]: And, btw, `sum / count` is the same as `mean` :)

Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Fair enough :)

File: rectools/metrics/popularity.py (hunk 11 above)
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Done.

File: rectools/metrics/popularity.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+from rectools.utils import select_by_type
+
+
+class AvgRecPopularity(MetricAtK):
+    r"""
+    Average Recommendations Popularity metric.
+
+    Calculate the average popularity of the recommended items in each list,
+    where "popularity" of item is the average number of ratings (interactions)
+    for this item.
```
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Docstring was changed.

File: rectools/metrics/popularity.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+    .. math::
+        ARP@k = \frac{1}{\left|U_{t}\right|}\sum_{u\in U_{t}^{}}\frac{\sum_{i\in L_{u}}\phi (i)}{\left| L_{u} \right |}
+
+    where
+    :math:`\phi (i)` is the number of times item i has been rated in the training set.
```
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Docstring was changed.

File: rectools/metrics/popularity.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    :math:`|U_{t}|` is the number of users in the test set.
+    :math:`L_{u}` is the list of recommended items for user u.
```
[the remark this answers is not in the document]
Comment (mikesokolovv) [reply rectools/metrics/popularity.py]: Docstring was changed.

**Review by feldlime — APPROVED** [review]

**Review by blondered — APPROVED** [review]
